medvis/dicom/src/io/gdcmvolumereader.cpp

```diff
@@ -0,0 +1,1070 @@
+           << createLine("Name:", p.patientName, 2) << createLine("ID", p.patientId, 2)
+           << createLine("No. Studies:", p.studies.size(), 2);
+        for (const auto& s : p.studies) {
+            ss << printStudy(s, 2);
+        }
+        ss << "]";
+
+        return ss.str();
+    };
+
+    for (const auto& p : dataPerPatient) {
+        LogInfoCustom("GdcmVolumeReader", printPatient(p));
+    }
+
+    return outputVolumes;
+}
+
+/**
+ * Entry point of the reader, called from VolumeSource processor
+ */
+SharedVolumeSequence GdcmVolumeReader::readData(const std::string& filePath) {
+    auto path = filePath;
+    if (!filesystem::fileExists(path)) {
+        std::string newPath = filesystem::addBasePath(path);
+        if (filesystem::fileExists(newPath)) {
+            path = newPath;
+        } else {
+            throw DataReaderException("Error could not find input file: " + path);
+        }
+    }
+
+    const auto directory = filesystem::getFileDirectory(path);
+    // TODO sequence source calls here for each file in a folder
+    if (directory == this->file_) {
+        return this->volumes_;  // doesnt work
+    }
+    gdcm::Trace::DebugOff();  // prevent gdcm from spamming inviwo console
+
+    // Try several dicom file structures
+
+    {
+        SharedVolumeSequence outputVolumes = tryReadDICOMDIR(directory);
+        if (outputVolumes) {
+            // Return now if reading was already successful
+            this->file_ = directory;
+            this->volumes_ = outputVolumes;
+            return outputVolumes;
+        }
+    }
+
+    {
+        SharedVolumeSequence outputVolumes = tryReadDICOMsequenceRecursive(directory);
+        if (outputVolumes) {
+            this->file_ = directory;
+            this->volumes_ = outputVolumes;
+            return outputVolumes;
+        }
+    }
+
+    // TODO read DICOM image file and detect the series it belongs to
+
+    // TODO make different dimensions possible in one sequence
+
+    // Otherwise keep trying
+    this->file_ = path;
+    SharedVolumeSequence outputVolumes = std::make_shared<VolumeSequence>();
+    gdcm::ImageReader reader;
+    reader.SetFileName(file_.c_str());
+    if (!reader.Read()) {
+        MevisVolumeReader mvreader;
+        if (mvreader.setFilenames(file_)) {
+            LogInfo("This seems to be a MevisLab dcm/tif file - calling the appropriate reader...");
+            SharedVolume v = mvreader.readData(file_);
+            outputVolumes->push_back(v);
+            return outputVolumes;
+        } else {
+            throw DataReaderException("Error could not read input file.");
+        }
+    }
+
+    const gdcm::Image& image = reader.GetImage();
+    const gdcm::File& file = reader.GetFile();
+    SharedVolume v = generateVolume(image, file);
+    if (!v) {
+        throw DataReaderException("Error could not read input file.");
+    }
+
+    outputVolumes->push_back(v);
+
+    return outputVolumes;
+}
+
+/**
```

Comment:
Ok this file needs a lot of work,,, but I suppose we will have to have this as is for now...

Reply:
Exactly.